### PR TITLE
implements a generic workload controller

### PR DESCRIPTION
**Pull request by p0lyn0mial**
it will be used by `openshift-apiserver` and `oauth-apiserver`

### CODE REVIEW

**Comment by p0lyn0mial** [comment]
I wired it manually in https://github.com/openshift/cluster-openshift-apiserver-operator/pull/333

File: pkg/operator/apiserver/controller/workload/workload.go [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+	}
+
+	workload, operatorConfigAtHighestGeneration, errs := c.syncFn()
+
+	return c.updateOperatorStatus(workload, operatorConfigAtHighestGeneration, errs)
+}
+
+// Run starts workload controller and blocks until stopCh is closed.
+// Note that setting workers doesn't have any effect, the controller is single-threaded.
+func (c *Controller) Run(ctx context.Context, workers int) {
```
Comment (p0lyn0mial) [inline pkg/operator/apiserver/controller/workload/workload.go]: Probably we should drop `workers` param

File: pkg/operator/apiserver/controller/workload/workload.go [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+	operatorSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+
+	if run, err := c.shouldSync(operatorSpec); !run {
+		return err
+	}
+
+	if fulfilled, err := c.preconditionFulfilled(operatorSpec); !fulfilled {
```
Comment (p0lyn0mial) [inline pkg/operator/apiserver/controller/workload/workload.go]: I think I am going to add a custom function for `precoditions`, for example when a config hasn't been populated there is no need to run the `sync` function. That function would be provided by the callers.

File: pkg/operator/apiserver/controller/workload/workload.go [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+
+// AddInformer queues the given informer to check spec, status and managed resources
+func (c *Controller) AddInformer(informer cache.SharedIndexInformer) *Controller {
+	informer.AddEventHandler(c.eventHandler())
+	c.preRunCachesSynced = append(c.preRunCachesSynced, informer.HasSynced)
+	return c
+}
+
+// AddNamespaceInformer queues the given ns informer for the targetNamespace
+func (c *Controller) AddNamespaceInformer(informer cache.SharedIndexInformer) *Controller {
```
Comment (sttts) [inline pkg/operator/apiserver/controller/workload/workload.go]: how odd that we have a special namespace informer code path. @mfojtik do you have that on the radar?

**Pull request merged**